src/hpr_routing.erl

```diff
@@ -155,35 +164,51 @@ packet_type_check(Packet) ->
         {uplink, _} -> true
     end.
 
--spec mic_check(Packet :: hpr_packet_up:packet()) -> boolean().
+-spec execute_checks(Packet :: hpr_packet_up:packet(), [{fun(), any()}]) -> ok | {error, any()}.
+execute_checks(_Packet, []) ->
+    ok;
+execute_checks(Packet, [{Fun, ErrorReason} | Rest]) ->
+    case Fun(Packet) of
+        false ->
+            {error, ErrorReason};
+        true ->
+            execute_checks(Packet, Rest)
+    end.
+
+-spec mic_check(Packet :: hpr_packet_up:packet()) -> boolean() | {true, [hpr_route:route()]}.
 mic_check(Packet) ->
     case hpr_packet_up:type(Packet) of
         {undefined, _} ->
             true;
         {join_req, _} ->
             true;
         {uplink, {_Type, DevAddr}} ->
-            case hpr_skf_ets:lookup_devaddr(DevAddr) of
-                {error, not_found} ->
+            case hpr_route_ets:lookup_skf(DevAddr) of
+                [] ->
                     true;
-                {ok, Keys} ->
+                SKFs ->
                     Payload = hpr_packet_up:payload(Packet),
-                    lists:any(
-                        fun(Key) ->
-                            hpr_lorawan:key_matches_mic(Key, Payload)
+                    Filtered = lists:filtermap(
+                        fun({Key, RouteID}) ->
+                            case hpr_lorawan:key_matches_mic(Key, Payload) of
```

Comment:
This mic_check will not consider routing for routes with no SKF if _any_ other Route with an overlapping devaddr range has an SKF.